### ISSUE TITLE
PDBs using minUnavailable prevent draining unready pods 

### ISSUE THREAD

**Issue by jhorwit2** [kubernetes/kubernetes#80389]
**What happened**:

We were unable to drain pods on an unhealthy node that were stuck in pod initializing. The disruption budget has a `minAvailable` of 2. We had two healthy pods and a pod stuck in initializing thus it's unhealthy. A drain on the node with the unhealthy pod stuck in initializing was prevented. 

**What you expected to happen**:

The pods should have been evicted because they are in a not Ready state. 

**How to reproduce it (as minimally and precisely as possible)**:

Using a PDB with `minAvailable` attempt to drain a node with a pod in a not Ready state. 

**Anything else we need to know?**:

The reason this is prevented is because the disruption controller determines the number of allowed disruptions by looking at [`currentHealthy - desiredHealthy`](https://github.com/kubernetes/kubernetes/blob/a07b027261383127494fc6d4e3ca2c544fb4a464/pkg/controller/disruption/disruption.go#L749-L752). When `minAvailable` is used your `desiredHealthy` count becomes that value, which in our case is 2. The `currentHealthy` count is 2 also because the 3rd pod is stuck in an initializing state on the bad node. This leaves the PDB with an allowed disruption of 0 and thus blocks the drain. 

The eviction logic in the storage only allows pods in a terminal state to bypass the PDB. 

https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/pod/storage/eviction.go#L124-L133

If they're not in a terminal state the eviction will only be allowed to occur if there are allowable disruptions available on the PDB.

It seems that the storage logic should allow draining pods that are not ready instead of just allowing terminal pods. 

/sig apps

**Comment by tedyu** [comment]
Are you suggesting that PodPending phase be handled in a similar manner to how PodSucceeded and PodFailed are handled ?

**Comment by michaelgugino** [comment]
I have run into this too, today.  I agree with the assessment of the situation and the proposed remedy.

**Comment by michaelgugino** [comment]
I'm still interested in this.  I'm hoping the ignore-pending case will merge at some point, and then we can investigate this.

I think between how we're doing deletion of pending (enforcing deletionOptions that ensure pod has not changed revision) and my other patch that checks current PDB count (not include the broken pod) is stable could work.

Since 'NotReady' is possibly transitive, we would need to ensure we always verify PDBs before removing a pod in this state.  It might be 'NotReady' because it's replicating data from another pod, etc, and removing it without checking there are enough healthy replicas could cause data loss, IMO.

**Comment by johngmyers** [comment]
If it's not ready because it's replicating data from another pod then the data exists in the pod it is replicating from.

If a pod is not ready then it does not contribute to the satisfaction of the PodDisruptionBudget constraint and deleting it will not cause that constraint to be violated.

**Comment by michaelgugino** [comment]
How about this scenario:

I have 3 pods.  1 pod hit some kind of problem, and the process exits.  The process takes some minutes to warm up, but all data is present.  This pod is 'NotReady' but otherwise working fine.  Simultaneously, another pod on a different node dies because the node crashes, never to return.

If we disregard PDBs in this scenario, we risk data loss by terminating the NotReady pod without regard to PDBs.

I think in this scenario, even if we check PDBs, we still might violate due to a race condition.

I'm not sure we can safely do this compared to the behavior we have today, at least not for the 'Not Ready' condition or any condition that indicates the pod is in 'Running' phase.  All we can say for certain when a pod is not ready is that it's not service traffic behind a to a service.  It could be do any number of things.

We might be able to tackle this on the drain client side with additional filters and options.  For instance, if a Pod is stuck in some non-Ready state, and has a lastTransitionTime > some configurable time, we could force delete that pod.

**Comment by johngmyers** [comment]
I would say the proposed fix is worse than the problem. It can take a long time for a database to replicate and a too-short timeout on forced deletion will lead to data loss.

A replication factor of two is vulnerable to two failures happening within the time it takes to replicate. That is fundamental.

I do believe you're conflating the ability to handle requests with ability for a database to handle disruption. For our StatefulSets we have the data pods indicate readiness on ability to handle requests and then have a Deployment of pods that indicate readiness on the ability of the database to tolerate loss of data pods. The PodDisruptionBudget then includes both StatefulSet and Deployment in its selector.

**Comment by michaelgugino** [comment]
So, I disagree.  Allowing it to be ignored (or force-deleted) client-side with a configurable timeout (say, 24hr last transition) results in the same thing as ignoring PDBs for this purpose:  the pod goes away.  Client-side fix is not worse than the problem, especially considering we can make the use opt-in (not default), and the timeout is configurable.  It's up to the user to decide when or if they wish to use that option.

Unfortunately, we can't assume anything about an application based on PDBs alone.  While some users might be utilizing pod readiness and PDBs in the manner you describe, probably not all are.

> For our StatefulSets we have the data pods indicate readiness on ability to handle requests 

In this scenario, a pod could have valid data that will be utilized for quorum once it rejoins, but is not yet ready.  It's not safe to ignore PDBs here if a pod is not ready, because we don't know if was previously ready, or just finished replicating all data from another (now gone) pod and is about to become ready, or what.  The pod could immediately transition to ready at any time and fix or keep an application from breaking.

Context, is of course, everything here.  The node in question was unhealthy (I presume unreachable is the more definitive state).  The eviction API doesn't consider what state the node is in.  The client doing the eviction can.

The context is also, IMO, automated systems trying to drain an unhealthy node.  We've already made the determination that the node is unhealthy and we want to remove workloads.  Most likely, that node is about to go away.  If it's just an admin running commands from the CLI, it's simple enough to delete a stuck pod and re-run drain.  An automated system can't really do this.

The tradeoff here is between never violating PDBs no matter what (current behavior of drain/eviction) or sometimes ignoring them.

**Comment by praseodym** [comment]
@michaelgugino:
> I have 3 pods. 1 pod hit some kind of problem, and the process exits. The process takes some minutes to warm up, but all data is present. This pod is 'NotReady' but otherwise working fine. Simultaneously, another pod on a different node dies because the node crashes, never to return.
> 
> If we disregard PDBs in this scenario, we risk data loss by terminating the NotReady pod without regard to PDBs.

Does the current drain/eviction procedure code handle this scenario at all? Let's say I have a deployment with `replicas: 3` and a PDB with `minUnavailable: 2`. A node gets drained so one pod gets evicted, then before a new pod is fully started another pod crashes. This leaves just one available pod, violating the PDB.

**Comment by michaelgugino** [comment]
> A node gets drained so one pod gets evicted, then before a new pod is fully started another pod crashes.

This would be an involuntary disruption, which PDBs can't defend against.  There are some other issues with PDBs that make this situation more complex as far as preserving the crashed node for any DR purposes, but there is always a window of time where data loss can occur.

**Comment by johngmyers** [comment]
> The tradeoff here is between never violating PDBs no matter what (current behavior of drain/eviction) or sometimes ignoring them.

If, as in the OP, there is a PDB with `minAvailable: 2` with two ready pods and one non-ready pod, how does evicting the non-ready pod cause the PDB to be violated? There are still two available pods.

If the pod were healthy it would have been evicted.

**Comment by michaelgugino** [comment]
> > The tradeoff here is between never violating PDBs no matter what (current behavior of drain/eviction) or sometimes ignoring them.
> 
> If, as in the OP, there is a PDB with `minAvailable: 2` with two ready pods and one non-ready pod, how does evicting the non-ready pod cause the PDB to be violated? There are still two available pods.
> 
> If the pod were healthy it would have been evicted.

Yeah, as long as we're validating we have the minAvailable, I think we could be okay.  There are opportunities for race conditions, but those would be involuntary disruptions and aren't covered in the code as is anyway.

**Comment by michaelgugino** [comment]
@liggitt could you reopen this?  I don't have permissions, there are other issues outstanding related to this bug.

**Comment by johngmyers** [comment]
/close
Looks like there was a commit merged fixing this. @michaelgugino please explain what the other issues are or open new issues for them.

**Comment by michaelgugino** [comment]
@johngmyers Yeah, that was my PR, it only covered the case of pending pods, not pods in any other state.

Per all the previous discussions here, I think we should aim to include all not ready pods.  The workflow is a bit different than what we did for just pending pods; for just pods stuck in pending, we know they were never ready, so we don't even have to check PDBs.  For pods that are not ready, we need to confirm that are enough pods that are actually available (according to PDB calculations) to be extra sure we're not going to delete something we shouldn't.  I have started some of the work for that previously here: https://github.com/kubernetes/kubernetes/pull/81175  But that patch was created before the one that merged, so needs a rebase and some re-design.
